Surprise

rgangela99 opened the pull request:
Implemented transformer to calculate surprise across a corpus based on some context. Also used this new transformer to reimplement the SpeakerConvoDiversity transformer.
There is a demo implemented in surprise_demo.ipynb that uses the surprise transformer to calculate speaker convo diversity, compares the results to those of the SpeakerConvoDiversity transformer and also compares the old SpeakerConvoDiversity transformer to the reimplemented version.

tisjune commented:
Thank you for your hard work, and for wrangling with the many parameters that could go into these functions!

Just to clarify: which of the demos are public, versus are for our benefit? It seems like the Tennis demo might be there to demo surprise, while the Reddit demo is more for us to verify that surprise == speakerconvodiversity? (or is it also to demo speakerconvodiversity?)

That clarification aside, from a first glance, I had some mostly cosmetic comments relating to the demos/documentation; there's one suggestion for the code. 

Documentation:
* https://github.com/CornellNLP/Cornell-Conversational-Analysis-Toolkit/blob/f6d60e02c14baa8be68b66a715feb700d1f05976/convokit/surprise/surprise.py#L50  maybe clarify what exactly "target" is, in a more interpretable/plain-English way -- really it's just an utterance that any other transformer would act on, right? 
* https://github.com/CornellNLP/Cornell-Conversational-Analysis-Toolkit/blob/f6d60e02c14baa8be68b66a715feb700d1f05976/convokit/surprise/surprise.py#L86 I don't see a "cv" parameter in the Surprise constructor, maybe I'm missing something? 

Code: 
* It's a little clunky that depending on the group_and_models setting, you end up with a ton of differently-named fields that correspond to the same property (e.g., see the tennis demo, where each utterance is its own group). A possible suggestion: have a special keyword such that if you pass in group_and_models = lambda utt: ('self', [model1, ...]), the metadata field will becalled GROUP_self__MODEL instead of GROUP_<individual id>_MODEL etc. Another suggestion, potentially allow the user to optionally specify a custom name that supercedes the group_and_models naming scheme.

Demos:
* replace all mentions of your own home directory (/home/axl4). You can look at other demos (e.g., PromptTypes) for how to do this -- basically, say something like path = <YOUR DIRECTORY HERE> such that the person running the demo will fill it out for themselves. Likewise, remove print outputs to your home directory.
* I don't think you need to print out entire dataframes (unless you think it's important that the person reading the demo sees everything that's in the dataframe -- so either print out aggregated stats, or print out specific columns and rows, for instance)
* perhaps depending on the answer to the clarification question above, maybe add a brief explanation of what surprise actually is, e.g., "this is based on the Tennis paper, we compare utterances to a language model using cross entropy, as implemented by surprise..."

Tennis demo:
* for part 1, cells 14/15, maybe provide a brief explanation of whether or not you expect these results (what do the numbers mean?) a statistical significance test wouldn't hurt either (if you want to rigorously check that the surprise for group x > surprise for group y)
* for part 2:
	* there's a typo in the section header ("posed ot female players")
	* mention at the _very start_ of the demo that you're adding player gender, because you want to make a comparison later on
	* similar to my part 1 comment, maybe provide a brief explanation of what conclusion we should be trying to draw, if any, from cells 15 and 16. (even if just a sanity check, i.e., "unsurprisingly female player questions sound like the female language model" -- answer the question posed by the section header)

Reddit demo:
* it would help to interpret the output a little more throughout, e.g., in cell 16, what does USERNAME_id mean? clarify that this is indexing into a speaker and a particular conversation. also, is there anything you can say to make sense of the output? e.g., perhaps it's plausible that speakers are going to be surprising/unsurprising across multiple convos, so seeing the same names show up again and again makes sense.

rgangela99 commented:
Thanks for the comments Justine!

Regarding the clarification on demos: I think the reddit demo could also be useful as a public demo to give another example of how surprise can be used. Maybe it would be better to make a cleaned up public version of the reddit demo so that it's just demoing a usage of surprise rather than comparing surprise to speakerconvodiversity.

Documentation: I added some clarification for the input "target". Let me know if you think it's enough. The "cv" parameter was leftover from a previous iteration. I removed it from the documentation.

Code: I added a parameter to the `transform` function to allow people to customize the keys used in the surprise metadata field. I'll update the tennis demo to demo how to use the new parameter and address the comments on the demos later.

tisjune commented:
Thanks for the updates! I think your proposed changes to the reddit and
tennis demos sound good -- in particular, having an explanation of
groupname and modelkey in a demo would help to clarify things even more, as
you said.  Once you're done with those demos then everything should be good
to go!

On Sat, May 15, 2021 at 4:43 PM rgangela99 ***@***.***> wrote:

> Thanks for the comments Justine!
>
> Regarding the clarification on demos: I think the reddit demo could also
> be useful as a public demo to give another example of how surprise can be
> used. Maybe it would be better to make a cleaned up public version of the
> reddit demo so that it's just demoing a usage of surprise rather than
> comparing surprise to speakerconvodiversity.
>
> Documentation: I added some clarification for the input "target". Let me
> know if you think it's enough. The "cv" parameter was leftover from a
> previous iteration. I removed it from the documentation.
>
> Code: I added a parameter to the transform function to allow people to
> customize the keys used in the surprise metadata field. I'll update the
> tennis demo to demo how to use the new parameter and address the comments
> on the demos later.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/CornellNLP/Cornell-Conversational-Analysis-Toolkit/pull/94#issuecomment-841722513>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBDQOECE2AK3DHCY4Z2VULTN3MHDANCNFSM43GXYVNA>
> .
>


-- 
Justine Zhang

tisjune commented:
Thanks so much! I just had some writing comments here and there. once you've addressed those (note there's a few that you can make the call to not address at all) let me know, and we can merge this branch! 


In general, try to  remove all the printed outputs that look like "Dataset already exists at <directory>" -- you can simply clear the output of that cell.

## surprise_demo:



### Step 2

typo: tokenizer will be an identi*ty function

> If the target or context is too short, the transformer will set the surprise to be nan. If we sent these to simply be 1, the most surprising statements tend to just be the very short statements.

to clarify == if we set the sample size to 1, right? (maybe just swap the order of these sentences so it's clearer)

brief intuition of why or why not smooth parameter? i.e., would we ever not use laplace smoothing?

### Step 4: 

* where's step 3?

* what does the transformer do if you call with a diferent object type? (might be helpful to reference the tennis demo here)

### analysis

briefly describe the output: what does <username>_<convoid> mean? briefly interpret too: does it make sense that the same users seem to show up over and over again? maybe output examples of what they contributed in a conversation (just try it out, if the output is really hard to interpret then feel free to not include that)

---


## tennis_demo:

briefly describe what the dataset is -- it's not game language so much as post-game interviews; also, the player gender wasn't inferred by us, it probably came from knowing whether it was the womens or mens games being played. you can lift language from the Tie-Breaker to do this explaining for you.

### part 1

do you need to select the random subset to compute surprise for? (iirc this is a fairly small dataset, but i might be wrong -- your call on what takes a reasonable amount of time)

when you compare female vs male players, could you show some sort of histogram, or do some sort of statistical significance test? (mann whitney u would work here)

briefly describe what it means for the average surprise score to be higher. in this context it means something like "female players are asked less relevant-to-tennis questions", right?

### part 2

you could make the distinction between part 1 and part 2 clearer: in part 1 we were comparing language to ALL questions, here we are training separate models for F and M players, right?

results: something (maybe obvious) to state outright is that surprise is lower when you compare the same gender to itself, vs when you compare questions asked to the opposite gender of what the langauge model was trained on, right? just saying that would help the reader make sense of what the surprise score is doing.

rgangela99 commented:
Thanks for the additional suggestions! I'm currently on a graduation trip and don't have my laptop with me, but I'll address these last comments once I'm home in a week and a half

tisjune commented:
sure, take your time, and enjoy your trip!

On Tue, Jun 15, 2021 at 3:37 AM rgangela99 ***@***.***> wrote:

> Thanks for the additional suggestions! I'm currently on a graduation trip
> and don't have my laptop with me, but I'll address these last comments once
> I'm home in a week and a half
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/CornellNLP/Cornell-Conversational-Analysis-Toolkit/pull/94#issuecomment-861261614>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBDQOHUEMB7EGHPYPHUVQ3TS37MNANCNFSM43GXYVNA>
> .
>


-- 
Justine Zhang

rgangela99 commented:
Hey Justine! I'm working on responding to your comments. There are a few that I wanted to respond to/ask about.

> ## surprise_demo:
> brief intuition of why or why not smooth parameter? i.e., would we ever not use laplace smoothing?

The reason for this was that we were originally modeling the surprise transformer after the speakerconvodiversity and that transformer did not use laplace smoothing.

> ## tennis_demo:
> briefly describe what the dataset is -- it's not game language so much as post-game interviews; also, the player gender wasn't inferred by us, it probably came from knowing whether it was the womens or mens games being played. you can lift language from the Tie-Breaker to do this explaining for you.

The demo uses two separate datasets and I believe one of them is from game commentary. I added descriptions of both to the demo.

> ### part 1
> do you need to select the random subset to compute surprise for? (iirc this is a fairly small dataset, but i might be wrong -- your call on what takes a reasonable amount of time)

I found that without taking a random subset, the demo took too long to run. I think it may be because the game commentary corpus that we are using for the context is larger. If I make the context sample size smaller it will probably run faster.

> when you compare female vs male players, could you show some sort of histogram, or do some sort of statistical significance test? (mann whitney u would work here)

I just tried making histograms and it seems like there's not really a significant difference between the two genders actually. I'm not sure if it's because the samples I'm taking from the interview questions are too small. I also noticed that the tie breaker paper uses a bigram model and Kneser-Ney smoothing whereas Surprise uses unigrams and laplace smoothing. Do you think that could also be a reason for the differing results?

> ### part 2
> you could make the distinction between part 1 and part 2 clearer: in part 1 we were comparing language to ALL questions, here we are training separate models for F and M players, right?

I actually just realized I never updated the heading for part 1. Part 1 is actually comparing questions to the tennis commentary dataset rather than to other questions. I updated the heading but let me know if you think it still needs more clarification.

> 
> results: something (maybe obvious) to state outright is that surprise is lower when you compare the same gender to itself, vs when you compare questions asked to the opposite gender of what the langauge model was trained on, right? just saying that would help the reader make sense of what the surprise score is doing.

I think I added something saying this in the text just under results. Let me know if there's a better place to put it.

tisjune commented:
Smoothing parameter: I think in this case just emphasizing that using
laplace smoothing is the default; the other non-smoothed option produces
less good results because <reason> works.

Random subset: I'd imagine this is probably the reason why the results are
different. How long does the demo take to run if you use the whole dataset?
Could you maybe give people the option to run on the whole dataset? (e.g.,
have a parameter that runs on the entire dataset if it's set to True)

I don't see an updated notebook in the repo, but what you said re: part 2
seems fine.

On Mon, Jun 28, 2021 at 10:42 PM rgangela99 ***@***.***>
wrote:

> Hey Justine! I'm working on responding to your comments. There are a few
> that I wanted to respond to/ask about.
>
> surprise_demo:
>
> brief intuition of why or why not smooth parameter? i.e., would we ever
> not use laplace smoothing?
>
> The reason for this was that we were originally modeling the surprise
> transformer after the speakerconvodiversity and that transformer did not
> use laplace smoothing.
>
> tennis_demo:
>
> briefly describe what the dataset is -- it's not game language so much as
> post-game interviews; also, the player gender wasn't inferred by us, it
> probably came from knowing whether it was the womens or mens games being
> played. you can lift language from the Tie-Breaker to do this explaining
> for you.
>
> The demo uses two separate datasets and I believe one of them is from game
> commentary. I added descriptions of both to the demo.
>
> part 1
>
> do you need to select the random subset to compute surprise for? (iirc
> this is a fairly small dataset, but i might be wrong -- your call on what
> takes a reasonable amount of time)
>
> I found that without taking a random subset, the demo took too long to
> run. I think it may be because the game commentary corpus that we are using
> for the context is larger. If I make the context sample size smaller it
> will probably run faster.
>
> when you compare female vs male players, could you show some sort of
> histogram, or do some sort of statistical significance test? (mann whitney
> u would work here)
>
> I just tried making histograms and it seems like there's not really a
> significant difference between the two genders actually. I'm not sure if
> it's because the samples I'm taking from the interview questions are too
> small. I also noticed that the tie breaker paper uses a bigram model and
> Kneser-Ney smoothing whereas Surprise uses unigrams and laplace smoothing.
> Do you think that could also be a reason for the differing results?
>
> part 2
>
> you could make the distinction between part 1 and part 2 clearer: in part
> 1 we were comparing language to ALL questions, here we are training
> separate models for F and M players, right?
>
> I actually just realized I never updated the heading for part 1. Part 1 is
> actually comparing questions to the tennis commentary dataset rather than
> to other questions. I updated the heading but let me know if you think it
> still needs more clarification.
>
> results: something (maybe obvious) to state outright is that surprise is
> lower when you compare the same gender to itself, vs when you compare
> questions asked to the opposite gender of what the langauge model was
> trained on, right? just saying that would help the reader make sense of
> what the surprise score is doing.
>
> I think I added something saying this in the text just under results. Let
> me know if there's a better place to put it.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/CornellNLP/Cornell-Conversational-Analysis-Toolkit/pull/94#issuecomment-870185955>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBDQOCUAVQZEL3RXAYYPMTTVEXIPANCNFSM43GXYVNA>
> .
>


-- 
Justine Zhang

rgangela99 commented:
I tried running the tennis demo with the entire interviews dataset last night and actually ended up getting a higher median surprise score for male players than female players. I'm not sure if there's something I'm doing wrong and not realizing. Do you have any insights?

tisjune commented:
Hmm:

1. what does the entire distribution look like? e.g., dataframe.groupby('meta.player_gender')['meta.surprise'].describe(). you could also use stats.mannwhitneyu to see if the differences are statistically significant.
2. if you inspect particular utterances that are more/less surprising, do those make sense? (one thought is to set a min/max wordcount so you don't get pathological surprise scores for unusually long/short utterances, you might have to play around with this a little; you could also try feeding in additional utterances listed as low/high perplexity in the paper and see if those match up). if this looks wrong, then it might make sense to play around with some of the other parameters, e.g., sampling more than 10 words per target.

minor point, i don't remember if nltk's tokenize automatically lowercases. 

also just realized that especially if you're comparing to the game commentary, it's a little inefficient to resample the context text every iteration. I realize that having a fixed context would require changing the code by quite a bit, so might not be necessary.

rgangela99 commented:
I'll look into what the distribution looks like and whether surprising/unsurprising utterances make sense.

nltk's tokenize doesn't automatically lowercase, but I tried running it with lowercase and removing punctuation and still got different results from the paper.

I was also thinking about whether sampling the context makes sense since all utterances are compared to the same context in the tennis demo. Ideally, we would want to just use the entire game language corpus as the context for each utterance, but currently, the surprise transformer requires the user to input a sample size. It cannot just select the entire context. I was thinking maybe we could allow `None` as an option for sample sizes in which case the entire text would be used and no sampling would occur. Do you think that makes sense?

tisjune commented:
yeah I think that makes total sense. just be sure to document it as well.

On Thu, Jul 1, 2021, 12:50 rgangela99 ***@***.***> wrote:

> I'll look into what the distribution looks like and whether
> surprising/unsurprising utterances make sense.
>
> nltk's tokenize doesn't automatically lowercase, but I tried running it
> with lowercase and removing punctuation and still got different results
> from the paper.
>
> I was also thinking about whether sampling the context makes sense since
> all utterances are compared to the same context in the tennis demo.
> Ideally, we would want to just use the entire game language corpus as the
> context for each utterance, but currently, the surprise transformer
> requires the user to input a sample size. It cannot just select the entire
> context. I was thinking maybe we could allow None as an option for sample
> sizes in which case the entire text would be used and no sampling would
> occur. Do you think that makes sense?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/CornellNLP/Cornell-Conversational-Analysis-Toolkit/pull/94#issuecomment-872401083>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBDQOB7O3EOW5NNDW4WWNLTVSMGLANCNFSM43GXYVNA>
> .
>

rgangela99 commented:
I did some qualitative analysis of the most and least surprising questions and they seem to make sense. The least surprising questions for both genders seem focused on tennis. Some examples of the most surprising questions: 

- Both you and Caroline have Polish origins. Do you speak Polish to each other?
- At the corporate luncheon, Annabel Croft made some unflattering remarks about your physique. Did that bother you or anything?
- Djokovic said his dog Pierre has gotten healthy since he went nongluten. Would consider the same for Bruno?
- Dumb question. Does anybody ask you about doing a socks commercial, sock endorsement?
- You heard from him today via text? He texted you today?

They seem overall to be not very related to tennis.

rgangela99 commented:
Hey Justine! I pushed some updates to the demos based on your comments and also the change to the surprise transformer to allow for the entire target/context to be used instead of taking samples. Let me know if it looks okay!

tisjune commented:
This looks good, thanks! One quick last thing to do is to fix the conflicting files -- the best thing to do is probably to use the current versions of __init__.py and setup.py in the master branch, and then add your code into those files.

rgangela99 commented:
Conflicts should be fixed now! Am I good to merge everything?

tisjune commented:
yep, go ahead, thanks!

calebchiam commented:
@all-contributors please add @rgangela99 for code
